org.scala-ide.sdt.core.tests/src/org/scalaide/core/testsetup/Bdd.scala

```diff
@@ -0,0 +1,71 @@
+package org.scalaide.core.testsetup
+
+import java.io.File
+import java.util.concurrent.locks.ReentrantLock
+
+import org.eclipse.core.resources.IProject
+import org.eclipse.core.resources.IncrementalProjectBuilder
+import org.eclipse.core.runtime.NullProgressMonitor
+import org.eclipse.jdt.core.IJavaProject
+import org.junit.Assert
+import org.scalaide.core.IScalaProject
+import org.scalaide.util.eclipse.EclipseUtils
+
+import SDTTestUtils.enableAutoBuild
+import SDTTestUtils.getErrorMessages
+import SDTTestUtils.sourceWorkspaceLoc
+import SDTTestUtils.workspace
+
+/**
+ * Behavioral Driven Development methods.
+ */
+trait Bdd {
+  import SDTTestUtils._
+
+  def givenCleanWorkspaceForProjects(projects: IProject*): Unit = {
+    workspace.build(IncrementalProjectBuilder.CLEAN_BUILD, new NullProgressMonitor)
+    projects.foreach { project =>
+      Assert.assertTrue(getErrorMessages(project).isEmpty)
+    }
+  }
+}
+
+object Bdd {
+  implicit def jProjectToIProject(project: IJavaProject): IProject = project.getProject
+  implicit def sProjectToIProject(project: IScalaProject): IProject = project.underlying
+}
+
+/**
+ * Before actions for Scala IDE tests.
+ */
+trait Before {
+  import SDTTestUtils._
+
+  private def withLock(lock: ReentrantLock)(f: => Unit) = try {
+    lock.lock()
+    f
+  } finally lock.unlock()
+
+  def initializeProjects(bundleName: String, projectNames: Seq[String])(postInit: => Unit): Unit = {
+    enableAutoBuild(false)
+    var done = false
+    val lock = new ReentrantLock()
+    val cond = lock.newCondition()
+    projectNames.foreach { name =>
+      EclipseUtils.workspaceRunnableIn(workspace) { monitor =>
+        withLock(lock) {
+          val wspaceLoc = workspace.getRoot.getLocation
+          val src = new File(sourceWorkspaceLoc(bundleName).toFile().getAbsolutePath + File.separatorChar + name)
+          val dst = new File(wspaceLoc.toFile().getAbsolutePath + File.separatorChar + name)
+          FileUtils.copyDirectory(src, dst)
+          done = true
+          cond.signal()
```

Comment:
This code looks fishy for several reasons.

- the code is copied from `org.scalaide.core.testsetup.SDTTestUtils.internalSetupProject(name: String, bundleName: String)`. Why? Can't we reuse the code there?
- the `workspaceRunnableIn` call could wrap the loop itself, so that all projects are copied in one go, instead of creating several transactions in a sequence. Is there a reason not to do that?
- the `await` call lower down will trigger after the first project in the list, but it doesn't seem to be the intention. Is that the intention?